File: leetcode/00253_meeting-rooms-ii/253-meeting-rooms-ii.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        memo = {}
        for _from, _to in intervals:
            memo[_from] = memo.get(_from, 0) + 1
            memo[_to] = memo.get(_to, 0) - 1

        memo = sorted(memo.items(), key=lambda kv: kv[0])
        curr_room = 0
        max_room = 0
        for _, diff in memo:
            curr_room += diff
            max_room = max(max_room, curr_room)
        return max_room

File: leetcode/00253_meeting-rooms-ii/253-meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        HashMap<Integer, Integer> memo = new HashMap<Integer, Integer>(0);
        for (int i = 0; i < intervals.length; i++) {
            memo.put(intervals[i][0], memo.getOrDefault(intervals[i][0], 0) + 1);
            memo.put(intervals[i][1], memo.getOrDefault(intervals[i][1], 0) - 1);
        }

        Integer max_room = 0;
        Integer curr_room = 0;
        SortedSet<Integer> keys = new TreeSet<>(memo.keySet());
        for (Integer key : keys) { 
            curr_room += memo.get(key);
            max_room = Math.max(max_room, curr_room);
        }
        return max_room;
    }
}
